### ISSUE TITLE
stub files and guava's Function

### ISSUE THREAD

**Issue by cushon** [typetools/checker-framework#559]
The declaration of guava's Function interface is:

```java
package com.google.common.base;

import javax.annotation.Nullable;

public interface Function<F, T> {
  @Nullable T apply(@Nullable F input);
}
```

The `@Nullable` annotations are there for findbugs, which interprets `@Nullable` to mean "only that some subtypes *might* accept null" ([guava/issues/920](https://github.com/google/guava/issues/920#issuecomment-61330676)). They clash with the checker's semantics, where the type of `apply` should just depend on the instantiation of the type parameters and you can have any of:

* `Function<@Nullable F, @Nullable T>`
* `Function<@Nullable F, @NonNull T>`
* `Function<@NonNull F, @Nullable T>`
* `Function<@NonNull F, @NonNull T>`

I can't find a way to write stubs that override the existing annotations.

---

Repro

Passing `null` to a `Function<@NonNull String, @NonNull Integer>` should result in an error, but succeeds anyways:

Function.astub

```java
package com.google.common.base;

import org.checkerframework.checker.nullness.qual.Nullable;

public interface Function<F extends @Nullable Object, T extends @Nullable Object> {
  T apply(F from);
}
```

Test.java

```java
import com.google.common.base.Function;
import org.checkerframework.checker.nullness.qual.NonNull;

class Test {
  void f(Function<@NonNull String, @NonNull Integer> x) {
    x.apply((String) null);
  }
}
```

```
$ javac -version -processor org.checkerframework.checker.nullness.NullnessChecker \
-cp guava.jar -Astubs=Function.astub Test.java
javac 1.8.0-jsr308-1.9.10
...
succeeds, should fail
```

---

Adding explicit annotations to `apply` in the stub works as expected, but the formal parameter needs to be un-annotated for it to depend on the instantiation of the type parameter.

Is there a way to write a stub that causes the analysis to ignore the existing annotation? Or is there some other approach you'd recommend?

**Comment by wmdietl** [comment]
Interesting problem. Removing the incorrect annotations would be a very simple solution, but I assume you don't want to do that :-)

There are four uses of field org.checkerframework.framework.stub.StubParser.atypeFactory in the StubParser that read the annotated type from the bytecode and then add the annotations from the stubfile.
In this situation, b/c the bytecode is already annotated and the stubfile contains no annotations, only the annotations from the bytecode persist.

One simple solution would be to add a new option "stubsOverrideBytecode" (better suggestions?) that removes the bytecode annotations before adding the stub annotations.
Maybe we can even make this the default: one usually uses stub files if the bytecode contains no annotations; if bytecode and stubfile contains annotations, they usually don't specify the same element; if bytecode and stubfile specify the same element, the stubfile wins, which might be useful, as illustrated by this issue.
So maybe we should make this changed behavior the default and add an option to warn about ignored annotations?

Anyone have other suggestions?

**Comment by mernst** [comment]
FindBugs's interpretation of `@Nullable` is confusing and unhelpful.  See the discussion at http://types.cs.washington.edu/checker-framework/current/checker-framework-manual.html#findbugs-nullable , which summarizes: "FindBugs has a non-standard definition of @Nullable. FindBugs’s treatment is not documented in its own Javadoc; it is different from the definition of @Nullable in every other tool for nullness analysis; it means the same thing as @NonNull when applied to a formal parameter; and it invariably surprises programmers."

Rather than degrade the quality of Guava's documentation by using FindBugs's `@Nullable` annotation (which in the referenced bug report and elsewhere has led to confusion), I think it would be better to properly annotate Guava with a `@Nullable` annotation that has defined, sensible semantics.
Another possibility would be to have a variant of Guava's jar file with Checker Framework annotations rather than FindBugs annotations.

If this is impossible, that would be unfortunate; a hack like Werner suggests might be necessary and ought to work.  A problem is that anyone reading the source code or Javadoc may become confused.  If "stubs override bytecode" is the default, then  perhaps the Checker Framework should print a warning whenever there both bytecode and stub file annotations exist for a given location.

**Comment by cushon** [comment]
We're in agreement about findbug's definition of `@Nullable`. I hope guava will migrate, but there's reluctance to do that until they can use type annotations without resorting to annotations-in-comments, and someone commits to migrating the entire codebase to use the annotations consistently. So unfortunately it's a bigger project than just deleting that one annotation.

The stub parser change Werner described sounds like the cleanest interim solution, and an improvement in general. But if it'd create other problems I won't lobby aggressively for it :)

Creating a separate jar to override the annotations has been discussed. I'd prefer to avoid yet another mechanism for overriding annotations, but it's an option.

> Maybe we can even make this the default: one usually uses stub files if the bytecode contains no annotations; if bytecode and stubfile contains annotations, they usually don't specify the same element; if bytecode and stubfile specify the same element, the stubfile wins, which might be useful, as illustrated by this issue.

Are the elements described here `javax.lang.model.element.Element` elements? I assume that creating a stub for one method in a class would only cause annotations present in bytecode to be ignored for that method, not the entire class?

**Comment by wmdietl** [comment]
The fix for this issue was merged today.
The -AstubWarnIfOverwritesBytecode option can be used to get messages when bytecode annotations are overwritten by a stub file.
Please let us know whether this fixes your issue.

**Comment by cushon** [comment]
I verified that the new stub handling and `-AstubWarnIfOverwritesBytecode` works perfectly for the `Function` stubs. Thank you!
